src/file_system/mod.rs

```diff
@@ -473,6 +494,64 @@ impl Directory {
         Directory::mkdir(Label::root(), Repo::repo_directory())
     }
 
+    /// Get the total size, in bytes, of a `Directory`. The size is
+    /// the sum of all files that can be reached from this `Directory`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use radicle_surf::file_system::{Directory, DirectoryContents, File, Label};
+    ///
+    /// let files = (
+    ///     DirectoryContents::file("main.rs".into(), b"println!(\"Hello, world!\")"),
+    ///     vec![DirectoryContents::file("lib.rs".into(), b"struct Hello(String)")],
+    /// ).into();
+    ///
+    /// let directory = Directory {
+    ///     label: Label::root(),
+    ///     entries: files,
+    /// };
+    ///
+    /// assert_eq!(directory.size(), 45);
+    /// ```
+    ///
+    /// ```
+    /// use nonempty::NonEmpty;
+    /// use radicle_surf::file_system::{Directory, DirectoryContents, File, Label};
+    ///
+    /// let mut entries: NonEmpty<DirectoryContents> = (
+    ///     DirectoryContents::file("main.rs".into(), b"println!(\"Hello, world!\")"),
+    ///     vec![DirectoryContents::file("lib.rs".into(), b"struct Hello(String)")],
+    /// ).into();
+    ///
+    /// let subdir = DirectoryContents::sub_directory(Directory {
+    ///     label: "test".into(),
+    ///     entries: NonEmpty::new(DirectoryContents::file(
+    ///         "mod.rs".into(),
+    ///         b"assert_eq!(1 + 1, 2);",
+    ///     )),
+    /// });
+    ///
+    /// entries.push(subdir);
+    ///
+    /// let directory = Directory {
+    ///     label: Label::root(),
+    ///     entries: entries,
+    /// };
+    ///
+    /// assert_eq!(directory.size(), 66);
+    /// ```
+    pub fn size(&self) -> usize {
```

Comment:
Hm, this can get very expensive. Do you think the `File` abstraction could be made such that it reads in the file contents lazily and just peeks at `blob.size()`?

Reply:
Just to clarify, do you mean it would be expensive to compute?

One thing I currently have is that I can grab the size from the `Blob` itself. I can push that commit in a second :)

Reply:
Yeah, you're reading the contents of all blobs into memory. It seems like [Blob::size()](https://docs.rs/git2/0.10.2/src/git2/blob.rs.html#38) and [Blob::content()](https://docs.rs/git2/0.10.2/src/git2/blob.rs.html#29) access the filesystem only when requested, so potentially `File` could delegate to the backend.

Probably good for now, but could use an optimisation pass later.

Reply:
Ah ok, so have `contents` and `size` be upon request rather than hold the data immediately. I'll create an issue for that and at least Application have the info to work with for the time being :)

Reply:
https://github.com/radicle-dev/radicle-surf/issues/17